crates/yielder/src/yo/client.rs

```diff
@@ -0,0 +1,317 @@
+        Self::deposit_call_data(vault.yo_token, assets, min_shares_out, receiver, partner_id)
+    }
+
+    pub fn redeem_call_data_for(vault: YoVault, shares: U256, min_assets_out: U256, receiver: Address, partner_id: u32) -> Vec<u8> {
+        Self::redeem_call_data(vault.yo_token, shares, min_assets_out, receiver, partner_id)
+    }
+
+    pub fn build_deposit_transaction(
+        &self,
+        from: Address,
+        yo_vault: Address,
+        assets: U256,
+        min_shares_out: U256,
+        receiver: Address,
+        partner_id: u32,
+    ) -> TransactionObject {
+        let data = Self::deposit_call_data(yo_vault, assets, min_shares_out, receiver, partner_id);
+        TransactionObject::new_call_with_from(&from.to_string(), &self.contract_address.to_string(), data)
+    }
+
+    pub fn build_redeem_transaction(
+        &self,
+        from: Address,
+        yo_vault: Address,
+        shares: U256,
+        min_assets_out: U256,
+        receiver: Address,
+        partner_id: u32,
+    ) -> TransactionObject {
+        let data = Self::redeem_call_data(yo_vault, shares, min_assets_out, receiver, partner_id);
+        TransactionObject::new_call_with_from(&from.to_string(), &self.contract_address.to_string(), data)
+    }
+
+    async fn call_gateway_contract<Call>(&self, call: Call) -> Result<Call::Return, YieldError>
+    where
+        Call: SolCall,
+    {
+        self.call_contract_at_block(call, self.contract_address, None).await
+    }
+
+    async fn call_contract_at_block<Call>(&self, call: Call, contract: Address, block_number: Option<u64>) -> Result<Call::Return, YieldError>
+    where
+        Call: SolCall,
+    {
+        let payload = hex::encode_prefixed(call.abi_encode());
+        let contract_address = contract.to_string();
+
+        let block_param = block_number
+            .map(|number| format!("0x{number:x}"))
+            .map_or_else(|| json!("latest"), serde_json::Value::String);
+
+        let response: String = self
+            .ethereum_client
+            .client
+            .call(
+                "eth_call",
+                json!([
+                    {
+                        "to": contract_address,
+                        "data": payload,
+                    },
+                    block_param
+                ]),
+            )
+            .await
+            .map_err(|err| YieldError::new(format!("yo gateway rpc call failed: {err}")))?;
+
+        if response.trim().is_empty() || response == "0x" {
+            return Err(YieldError::new("yo gateway response did not contain data"));
+        }
+
+        let decoded = hex::decode(&response).map_err(|err| YieldError::new(format!("invalid hex returned by yo gateway: {err}")))?;
+        Call::abi_decode_returns(&decoded).map_err(|err| YieldError::new(format!("failed to decode yo gateway response: {err}")))
+    }
+}
+
+#[async_trait]
+impl<C> YoGatewayApi for YoGatewayClient<C>
```

Comment:
```suggestion
impl<C> YoProvider for YoGatewayClient<C>
```

No need to use API in naming, consistent with swapper